fix: allow deep linking

kewde opened the pull request:
This PR removes the restriction of only allowing to open up "https" URLs, and instead allows the webview to open up "bitcoin:" deeplinks.

ChALkeR reviewed (dismissed):
nACK
We should create an allowlist of protocols (e.g. we don't want `http:`, `file:`, etc and basically all unvetted protocols)
Also, do we want `bitcoin:` for all websites?

633kh4ck commented:
> atm, (as i understood it) we are seeing that popup so the solution was to prevent it from displaying

@timlanahan atm, we don't allow (i.e., silently block) these deep links

633kh4ck commented:
> We would also need to make sure that insecure schemes such as http:// and file:// remain blocked.
> <...>
> I'd suggest to make the allowed schemes configurable from the app (similar to what we do with originWhitelist)

What if we combine both approaches? We could implement a denylist here (it must take precedence, so it can't be bypassed) and expose the setting to configure allowed schemes from the app land. This would provide a good balance between security and UX, IMO.

@matias-la @ChALkeR WDYT?

> Also, do we want bitcoin: for all websites?

Any specific concerns?

timlanahan commented:
Sounds good to me. Long term: can we control that UX with a custom component?  System prompts always feel scarier to me

633kh4ck commented:
> Long term: can we control that UX with a custom component?

No objections from a security PoV.

633kh4ck reviewed (commented):
@kewde Can you please also implement a hardcoded blocklist? We should block the `http`, `file`, and `javascript` protocols (before checking the allowed ones).

ChALkeR commented:
> (it must take precedence, so it can't be bypassed) and expose the setting to configure allowed schemes from the app land.

I'm afraid that might get degraded e.g. by using a `*` on the client side.

It might be fine if we
1) implement a denylist here, blocking eg. `file` and `http`
2) add an option for an allowlist of protocols, which should be full strings / no catch-all / no regexps
3) default that allowlist to `['https:']`

kewde commented:
> implement a denylist here, blocking eg. file and http (should use resolved protos)

I've implemented a blocklist, I've made sure to make it case insensitive, anything else?

https://github.com/ExodusMovement/react-native-webview/blob/b2563ba4efea95dad38a97b4c2badd1bc0a9e1fe/src/WebViewShared.tsx#L20


> add an option for an allowlist of protocols, which should be full strings / no catch-all / no regexps

I've added an option, it should use full strings. The assert for type string is done by the escape string regexp package.
![image](https://github.com/ExodusMovement/react-native-webview/assets/13770485/02fa7ecc-91bf-417d-8651-b54a74b565b0)


> default that allowlist to ['https:'] for the case when it's not specified

Done.

timlanahan commented:
if we're good here @matias-la @633kh4ck @ChALkeR we can merge and bump in mobile for audits.  does it sound okay?

633kh4ck reviewed (commented):
General feedback (can be addressed later, not a blocker)

633kh4ck reviewed (changes requested):
nACK

> The assert for type string is done by the escape string regexp package.

https://github.com/ExodusMovement/react-native-webview/blob/b2563ba4efea95dad38a97b4c2badd1bc0a9e1fe/src/WebViewShared.tsx#L69-L70

Security of `deepLinkWhitelist` can be degraded in the app land because of two facts:
 - the current implementation allows wildcards (due to `stringWhitelistToRegex`): https://github.com/ExodusMovement/react-native-webview/blob/b2563ba4efea95dad38a97b4c2badd1bc0a9e1fe/src/WebViewShared.tsx#L28
 - the current matcher uses regular expressions for checking against `deepLinkWhitelist`: https://github.com/ExodusMovement/react-native-webview/blob/b2563ba4efea95dad38a97b4c2badd1bc0a9e1fe/src/WebViewShared.tsx#L34
 
 Instead, we must use a matcher based on `String.prototype.startsWith()`.

kewde commented on src/WebViewShared.tsx:
```diff
@@ -47,22 +56,33 @@ const createOnShouldStartLoadWithRequest = (
-      });
+      const foundMatchInBlocklist = matchWithRegexList(defaultDeeplinkBlocklist, url)
+
+      const _deeplinkWhitelist = (deepLinkWhitelist || []).map(stringWhitelistToRegex)
```
I think this one is more fine since it would optimize the "sad" codepath that doesn't happen often, by slowing down the happy code path. It doesn't make sense to aalways precompute this whitelist when in reality we only need it 1% of the time.

633kh4ck replied:
Makes sense, I agree.

633kh4ck commented on src/WebViewShared.tsx:
```diff
@@ -16,29 +16,38 @@ import {
 import styles from './WebView.styles';
 
 const defaultOriginWhitelist = ['https://*'] as const;
+const defaultDeeplinkWhitelist = ['https://*'] as const;
```
Also, this is a bit confusing because the `https` protocol is not a deep link.

633kh4ck replied:
cc/ @ChALkeR

kewde replied:
> Also, this is a bit confusing because the https protocol is not a deep link.

In this context it is. `https:` is a registered deep link handler by Chrome or whatever app browser app.

633kh4ck replied:
I don't have a strong opinion here. I believe it's not a blocker for this iteration. @ChALkeR, please let me know if you think this requires revision.

kewde commented on src/WebViewShared.tsx:
```diff
@@ -16,29 +16,38 @@ import {
   `^${escapeStringRegexp(originWhitelist).replace(/\\\*/g, '.*')}$`;
 
+const matchWithRegexList = (
+  compiledWhitelist: readonly string[] | readonly RegExp[],
```
fixed in d6709fe8c640899499092dc616c0669f2a2c2f73

kewde commented on src/WebViewShared.tsx:
```diff
@@ -16,29 +16,38 @@ import {
 };
 
-const originWhitelistToRegex = (originWhitelist: string): string =>
+const stringWhitelistToRegex = (originWhitelist: string): string =>
```
fixed in d6709fe8c640899499092dc616c0669f2a2c2f73

kewde commented on src/WebViewShared.tsx:
```diff
@@ -47,22 +56,33 @@ const createOnShouldStartLoadWithRequest = (
+      const _deeplinkWhitelist = (deepLinkWhitelist || []).map(stringWhitelistToRegex)
+      const foundMatchInAllowlist = matchWithRegexList(_deeplinkWhitelist, url)
+
+      if (!foundMatchInBlocklist && foundMatchInAllowlist) {
```
fixed in [c3ef5b1](https://github.com/ExodusMovement/react-native-webview/pull/21/commits/c3ef5b127dc5fe4934c62807ae464f7f5af44d11)

kewde commented on src/WebViewShared.tsx:
```diff
@@ -47,22 +56,33 @@ const createOnShouldStartLoadWithRequest = (
     let shouldStart = true;
     const { url, lockIdentifier, isTopFrame } = nativeEvent;
 
     if (!_passesWhitelist(compileWhitelist(originWhitelist), url)) {
```
fixed in [c3ef5b1](https://github.com/ExodusMovement/react-native-webview/pull/21/commits/c3ef5b127dc5fe4934c62807ae464f7f5af44d11)

633kh4ck commented on src/WebViewShared.tsx:
```diff
@@ -34,6 +34,14 @@ const matchWithRegexList = (
+  value: string,
+) => {
+  if (typeof value !== 'string') throw new Error(`value was not a string`)
+  return prefixes.some(x => value.startsWith(x));
```
`String.prototype.startsWith` returns `true` if the given characters are found at the beginning of the string, including when `searchString` is an **empty string**; also, all `searchString` values that are not regexes are **coerced to strings** ([MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/startsWith)).

So
```js
prefixes.some(x => value.startsWith(x))
```
can be bypassed with
```js
prefixes = ['']
```
And
```js
prefixes.some(x => x.length && value.startsWith(x))
```
can be bypassed with
```js
prefixes = [['']]
```

---
Hardened implementation:
```suggestion
  return prefixes.some(x => String(x).length && String.prototype.startsWith.call(value, x));
```

633kh4ck replied:
The proposed impl. still suffers from weakness. The point is that without additional checks, it's still possible to get all deep links approved (`['a', 'b', 'c', 'd', 'e', 'f', 'u']`). @ChALkeR, thoughts on how to harden it further?

633kh4ck reviewed (changes requested):
nACK (the last one, I ~~promise~~ hope)

633kh4ck commented on src/WebViewShared.tsx:
```diff
@@ -16,29 +16,46 @@ import {
+  value: string,
+) => {
+  if (typeof value !== 'string') throw new Error(`value was not a string`)
+  return prefixes.some(x => String(x).length && String.prototype.startsWith.call(value, x));
```
Would be good to add tests for [these](https://github.com/ExodusMovement/react-native-webview/pull/21#discussion_r1619322987) cases.

633kh4ck reviewed (approved):
utACK

ChALkeR commented on src/WebViewShared.tsx:
```diff
@@ -47,22 +64,38 @@ const createOnShouldStartLoadWithRequest = (
+      const foundMatchInBlocklist = matchWithRegexList(defaultDeeplinkBlocklist, url)
+      if (!foundMatchInBlocklist) {
+        /** Check if the url passes the dynamic deeplink allow list */
+        const foundMatchInAllowlist = matchWithPrefixStringList(deepLinkWhitelist, url)
```
this implementation can fall through on `['']` being used as a whitelist

It should check something like `allowlist.includes(new URL(url).protocol)`

633kh4ck replied:
Addressed in https://github.com/ExodusMovement/react-native-webview/pull/23
